File: traka/lib/services/driver_nav_premium_pricing.dart
import '../config/driver_iap_catalog.dart';

/// Tarif navigasi premium: tier jarak (travel jauh) × pengali jenis rute, lalu snap ke SKU Play.
///
/// Firestore `app_config/settings` (opsional):
/// - [driverNavPremiumDistancePricingEnabled]: `false` = hanya tarif per-scope lama.
/// - [driverNavPremiumTierMaxKm]: daftar batas atas km per pita (naik), mis. [75,200,450,900,1500,1e9].
/// - [driverNavPremiumTierBaseFeesRupiah]: biaya dasar per pita (sama panjang dengan tier max km).
/// - [driverNavPremiumScopeMultBpsDalam] / …Antar / …Nasional: basis poin (100 = 1.00).
/// - [driverNavPremiumSnapFeesRupiah]: nominal IAP yang diizinkan (snap terdekat).
class DriverNavPremiumPricing {
  DriverNavPremiumPricing._();

  static const int minFeeRupiah = 9000;
  static const int maxFeeRupiah = 150000;

  /// Batas kepercayaan jarak dari klien (anti input ngawur).
  static const double maxTrustedDistanceMeters = 2500000;

  static const List<double> _defaultTierMaxKm = [
    75,
    200,
    450,
    900,
    1500,
    1e9,
  ];

  static const List<int> _defaultTierBaseFees = [
    10000,
    18000,
    28000,
    42000,
    55000,
    68000,
  ];

  /// Snap ke SKU yang sama dengan kontribusi driver ([kDriverDuesAmounts]) — satu set nominal Play.
  static List<int> get _defaultSnapFees => List<int>.from(kDriverDuesAmounts);

  /// Tarif lama: hanya [scope] + field Firestore tier tunggal (tanpa jarak).
  static int legacyScopeOnlyRupiah(String? scopeName, Map<String, dynamic>? d) {
    const min = 3000;
    int readPositive(String key) {
      final v = d?[key];
      if (v != null) {
        final n = (v is num) ? v.toInt() : int.tryParse(v.toString());
        if (n != null && n >= min) return n;
      }
      return 0;
    }

    int legacy() {
      final v = d?['driverNavPremiumFeeRupiah'];
      if (v != null) {
        final n = (v is num) ? v.toInt() : int.tryParse(v.toString());
        if (n != null && n >= min) return n;
      }
      return 0;
    }

    int forScope(String? scope) {
      switch (scope) {
        case 'dalamProvinsi':
          final x = readPositive('driverNavPremiumFeeDalamProvinsiRupiah');
          if (x > 0) return x;
          final l = legacy();
          if (l > 0) return l;
          return 50000;
        case 'antarProvinsi':
          final x = readPositive('driverNavPremiumFeeAntarProvinsiRupiah');
          if (x > 0) return x;
          final l = legacy();
          if (l > 0) return l;
          return 75000;
        case 'dalamNegara':
          final x = readPositive('driverNavPremiumFeeNasionalRupiah');
          if (x > 0) return x;
          final l = legacy();
          if (l > 0) return l;
          return 100000;
        default:
          final l = legacy();
          if (l > 0) return l;
          return 100000;
      }
    }

    return forScope(scopeName);
  }

  /// Satu angka final untuk IAP `traka_driver_nav_premium_<rupiah>`.
  static int computeRupiah({
    required String? scope,
    required double? distanceMeters,
    required Map<String, dynamic>? settings,
  }) {
    final enabled = settings?['driverNavPremiumDistancePricingEnabled'];
    final useDist = enabled != false &&
        distanceMeters != null &&
        distanceMeters > 0 &&
        distanceMeters <= maxTrustedDistanceMeters;

    if (!useDist) {
      return legacyScopeOnlyRupiah(scope, settings);
    }

    final km = distanceMeters / 1000.0;
    var maxKms = _parseTierMaxKm(settings);
    var bases = _parseTierBases(settings, maxKms.length);
    if (bases.length != maxKms.length) {
      maxKms = List<double>.from(_defaultTierMaxKm);
      bases = List<int>.from(_defaultTierBaseFees);
    }
    final snap = _parseSnapFees(settings);
    final base = _baseFeeForKm(km, maxKms, bases);
    final multBps = _scopeMultiplierBps(scope, settings);
    var rawInt = (base * multBps / 100.0).round();
    if (rawInt < minFeeRupiah) rawInt = minFeeRupiah;
    if (rawInt > maxFeeRupiah) rawInt = maxFeeRupiah;
    return _snapToNearest(rawInt, snap);
  }

  static List<double> _parseTierMaxKm(Map<String, dynamic>? d) {
    final raw = d?['driverNavPremiumTierMaxKm'];
    if (raw is! List || raw.length < 2) {
      return List<double>.from(_defaultTierMaxKm);
    }
    final out = <double>[];
    for (final e in raw) {
      final n = (e is num) ? e.toDouble() : double.tryParse(e.toString());
      if (n != null && n > 0) out.add(n);
    }
    if (out.length < 2) return List<double>.from(_defaultTierMaxKm);
    for (var i = 1; i < out.length; i++) {
      if (out[i] <= out[i - 1]) {
        return List<double>.from(_defaultTierMaxKm);
      }
    }
    return out;
  }

  static List<int> _parseTierBases(Map<String, dynamic>? d, int len) {
    final raw = d?['driverNavPremiumTierBaseFeesRupiah'];
    if (raw is! List || raw.length != len) {
      return len == _defaultTierMaxKm.length
          ? List<int>.from(_defaultTierBaseFees)
          : <int>[];
    }
    final out = <int>[];
    for (final e in raw) {
      final n = (e is num) ? e.toInt() : int.tryParse(e.toString());
      if (n == null || n < minFeeRupiah) {
        return len == _defaultTierMaxKm.length
            ? List<int>.from(_defaultTierBaseFees)
            : <int>[];
      }
      out.add(n);
    }
    return out;
  }

  static List<int> _parseSnapFees(Map<String, dynamic>? d) {
    final raw = d?['driverNavPremiumSnapFeesRupiah'];
    if (raw is! List || raw.isEmpty) {
      return List<int>.from(_defaultSnapFees);
    }
    final out = <int>[];
    for (final e in raw) {
      final n = (e is num) ? e.toInt() : int.tryParse(e.toString());
      if (n != null && n >= minFeeRupiah && n <= maxFeeRupiah) out.add(n);
    }
    out.sort();
    if (out.isEmpty) return List<int>.from(_defaultSnapFees);
    return out;
  }

  static int _scopeMultiplierBps(String? scope, Map<String, dynamic>? d) {
    int readBps(String key, int fallback) {
      final v = d?[key];
      if (v == null) return fallback;
      final n = (v is num) ? v.toInt() : int.tryParse(v.toString());
      if (n == null || n < 50 || n > 300) return fallback;
      return n;
    }

    switch (scope) {
      case 'dalamProvinsi':
        return readBps('driverNavPremiumScopeMultBpsDalam', 100);
      case 'antarProvinsi':
        return readBps('driverNavPremiumScopeMultBpsAntar', 108);
      case 'dalamNegara':
      default:
        return readBps('driverNavPremiumScopeMultBpsNasional', 116);
    }
  }

  static int _baseFeeForKm(double km, List<double> maxKms, List<int> bases) {
    for (var i = 0; i < maxKms.length && i < bases.length; i++) {
      if (km <= maxKms[i]) return bases[i];
    }
    return bases.isNotEmpty ? bases.last : _defaultTierBaseFees.last;
  }

  static int _snapToNearest(int raw, List<int> allowed) {
    if (allowed.isEmpty) {
      var x = (raw / 1000).round() * 1000;
      if (x < minFeeRupiah) x = minFeeRupiah;
      if (x > maxFeeRupiah) x = maxFeeRupiah;
      return x;
    }
    var best = allowed.first;
    var bestDiff = (raw - best).abs();
    for (final a in allowed) {
      final d = (raw - a).abs();
      if (d < bestDiff) {
        bestDiff = d;
        best = a;
      }
    }
    if (best < minFeeRupiah) return minFeeRupiah;
    if (best > maxFeeRupiah) return maxFeeRupiah;
    return best;
  }
}

File: traka/functions/lib/driverNavPremiumPricing.js
/**
 * Menyelaraskan tarif navigasi premium driver dengan Dart: DriverNavPremiumPricing.
 * Input settings = data app_config/settings.
 */

const minFeeRupiah = 9000;
const maxFeeRupiah = 150000;
const maxTrustedDistanceMeters = 2500000;

const defaultTierMaxKm = [75, 200, 450, 900, 1500, 1e9];
const defaultTierBaseFees = [10000, 18000, 28000, 42000, 55000, 68000];
const defaultSnapFees = [
  5000, 7500, 10000, 12500, 15000, 17500, 20000, 25000, 30000, 40000,
  50000, 60000, 75000, 100000, 150000, 200000,
];

function readPositive(d, key, min = 3000) {
  const v = d?.[key];
  if (v == null) return 0;
  const n = typeof v === "number" ? v : parseInt(String(v), 10);
  return !isNaN(n) && n >= min ? n : 0;
}

function legacyScopeOnlyRupiah(scopeName, d) {
  const min = 3000;
  function legacy() {
    const v = d?.driverNavPremiumFeeRupiah;
    if (v == null) return 0;
    const n = typeof v === "number" ? v : parseInt(String(v), 10);
    return !isNaN(n) && n >= min ? n : 0;
  }
  function forScope(scope) {
    switch (scope) {
      case "dalamProvinsi": {
        const x = readPositive(d, "driverNavPremiumFeeDalamProvinsiRupiah", min);
        if (x > 0) return x;
        const l = legacy();
        if (l > 0) return l;
        return 50000;
      }
      case "antarProvinsi": {
        const x = readPositive(d, "driverNavPremiumFeeAntarProvinsiRupiah", min);
        if (x > 0) return x;
        const l = legacy();
        if (l > 0) return l;
        return 75000;
      }
      case "dalamNegara": {
        const x = readPositive(d, "driverNavPremiumFeeNasionalRupiah", min);
        if (x > 0) return x;
        const l = legacy();
        if (l > 0) return l;
        return 100000;
      }
      default: {
        const l = legacy();
        if (l > 0) return l;
        return 100000;
      }
    }
  }
  return forScope(scopeName);
}

function parseTierMaxKm(d) {
  const raw = d?.driverNavPremiumTierMaxKm;
  if (!Array.isArray(raw) || raw.length < 2) return defaultTierMaxKm.slice();
  const out = [];
  for (const e of raw) {
    const n = typeof e === "number" ? e : parseFloat(String(e));
    if (!isNaN(n) && n > 0) out.push(n);
  }
  if (out.length < 2) return defaultTierMaxKm.slice();
  for (let i = 1; i < out.length; i++) {
    if (out[i] <= out[i - 1]) return defaultTierMaxKm.slice();
  }
  return out;
}

function parseTierBases(d, len) {
  const raw = d?.driverNavPremiumTierBaseFeesRupiah;
  if (!Array.isArray(raw) || raw.length !== len) {
    return len === defaultTierMaxKm.length ? defaultTierBaseFees.slice() : [];
  }
  const out = [];
  for (const e of raw) {
    const n = typeof e === "number" ? e : parseInt(String(e), 10);
    if (n == null || isNaN(n) || n < minFeeRupiah) {
      return len === defaultTierMaxKm.length ? defaultTierBaseFees.slice() : [];
    }
    out.push(n);
  }
  return out;
}

function parseSnapFees(d) {
  const raw = d?.driverNavPremiumSnapFeesRupiah;
  if (!Array.isArray(raw) || raw.length === 0) return defaultSnapFees.slice();
  const out = [];
  for (const e of raw) {
    const n = typeof e === "number" ? e : parseInt(String(e), 10);
    if (n != null && !isNaN(n) && n >= minFeeRupiah && n <= maxFeeRupiah) out.push(n);
  }
  out.sort((a, b) => a - b);
  if (out.length === 0) return defaultSnapFees.slice();
  return out;
}

function scopeMultiplierBps(scope, d) {
  function readBps(key, fallback) {
    const v = d?.[key];
    if (v == null) return fallback;
    const n = typeof v === "number" ? v : parseInt(String(v), 10);
    if (isNaN(n) || n < 50 || n > 300) return fallback;
    return n;
  }
  switch (scope) {
    case "dalamProvinsi":
      return readBps("driverNavPremiumScopeMultBpsDalam", 100);
    case "antarProvinsi":
      return readBps("driverNavPremiumScopeMultBpsAntar", 108);
    case "dalamNegara":
    default:
      return readBps("driverNavPremiumScopeMultBpsNasional", 116);
  }
}

function baseFeeForKm(km, maxKms, bases) {
  for (let i = 0; i < maxKms.length && i < bases.length; i++) {
    if (km <= maxKms[i]) return bases[i];
  }
  return bases.length > 0 ? bases[bases.length - 1] : defaultTierBaseFees[defaultTierBaseFees.length - 1];
}

function snapToNearest(raw, allowed) {
  if (!allowed || allowed.length === 0) {
    let x = Math.round(raw / 1000) * 1000;
    if (x < minFeeRupiah) x = minFeeRupiah;
    if (x > maxFeeRupiah) x = maxFeeRupiah;
    return x;
  }
  let best = allowed[0];
  let bestDiff = Math.abs(raw - best);
  for (const a of allowed) {
    const di = Math.abs(raw - a);
    if (di < bestDiff) {
      bestDiff = di;
      best = a;
    }
  }
  if (best < minFeeRupiah) return minFeeRupiah;
  if (best > maxFeeRupiah) return maxFeeRupiah;
  return best;
}

/**
 * @param {{ scope: string|null|undefined, distanceMeters: number|null|undefined, settings: object|null }} p
 * @return {number}
 */
function computeNavPremiumRupiah({ scope, distanceMeters, settings }) {
  const d = settings || {};
  const enabled = d.driverNavPremiumDistancePricingEnabled === true;
  const useDist = enabled &&
    distanceMeters != null &&
    !isNaN(distanceMeters) &&
    distanceMeters > 0 &&
    distanceMeters <= maxTrustedDistanceMeters;

  if (!useDist) {
    return legacyScopeOnlyRupiah(scope || "dalamNegara", d);
  }

  const km = distanceMeters / 1000.0;
  let maxKms = parseTierMaxKm(d);
  let bases = parseTierBases(d, maxKms.length);
  if (bases.length !== maxKms.length) {
    maxKms = defaultTierMaxKm.slice();
    bases = defaultTierBaseFees.slice();
  }
  const snap = parseSnapFees(d);
  const base = baseFeeForKm(km, maxKms, bases);
  const multBps = scopeMultiplierBps(scope || "dalamNegara", d);
  let rawInt = Math.round((base * multBps) / 100.0);
  if (rawInt < minFeeRupiah) rawInt = minFeeRupiah;
  if (rawInt > maxFeeRupiah) rawInt = maxFeeRupiah;
  return snapToNearest(rawInt, snap);
}

module.exports = {
  computeNavPremiumRupiah,
  minFeeRupiah,
  maxFeeRupiah,
};
